src/models/recommender.py

```diff
@@ -0,0 +1,208 @@
+
+
+    def predict(self, user_id, item_id):
+        '''docstring'''
+
+        try:
+            user_row = np.where(self.user_ids_series == user_id)[0][0]
+            item_col = np.where(self.item_ids_series == item_id)[0][0]
+
+            pred = np.dot(self.user_matrix[user_row, :],
+                          self.item_matrix[:, item_col])
+
+            return pred
+
+        except IndexError:
+            print("Prediction cannot be made because user_id or item_id does \
+            not exist in current database.")
+
+
+    def find_similar_items(self, input_id, items_df, key):
+        '''docstring'''
+
+        content = np.array(items_df.iloc[:, 3:])
+        items_dot_prod = content.dot(np.transpose(content))
+
+        # find the row of each movie id
+        item_idx = np.where(items_df[key] == input_id)[0][0]
+
+        # find the most similar movie indices
+        n_rec = self.n_rec
+        similar_idxs = np.argpartition(items_dot_prod[item_idx], -n_rec)[-n_rec:][::-1]
+
+        return similar_idxs
+
+
+    def find_popular_items(self, ranked_df):
+        '''docstring'''
+
+        n_rec = self.n_rec
+        top_items = list(ranked_df['name'][:n_rec])
+
+        return top_items
+
+
+    def make_recs(self, input_id):
+        '''docstring'''
+
+        n_rec = self.n_rec
+        rec_ids, rec_names = None, None
+
+        if input_id in self.user_ids_series:
+            idx = np.where(self.user_ids_series == input_id)[0][0]
+            preds = self.user_matrix[idx, :].dot(self.item_matrix)
+            indices = preds.argsort()[-n_rec:][::-1]
+            rec_ids = self.item_ids_series[indices]
+            rec_names = self._get_item_names(self.items, 'sku', rec_ids)
+
+        elif input_id in self.item_ids_series:
+            rec_ids = self.find_similar_items(input_id, self.items, 'sku')[:n_rec]
+            rec_names = self._get_item_names(rec_ids, self.items, 'sku')
+
+        else:
+            ranked_df = self._create_ranked_df()
+            rec_names = self.find_popular_items(ranked_df)
+
+        return rec_ids, rec_names
+
+
+    def _create_user_item_matrix(self, by=None, rating_col=None):
+        '''docstring'''
+
+        user_item_df = self.reviews.groupby(by)[rating_col].max().unstack()
+        self.user_ids_series = np.array(user_item_df.index)
+        self.item_ids_series = np.array(user_item_df.columns)
+
+        return np.array(user_item_df)
+
+
+    def _get_user_item_dimensions(self):
+        '''docstring'''
+
+        n_users = self.user_item_matrix.shape[0]
+        n_items = self.user_item_matrix.shape[1]
+        n_ratings = np.count_nonzero(~np.isnan(self.user_item_matrix))
+
+        return n_users, n_items, n_ratings
+
+
+    def _create_ranked_df(self):
+        '''docstring'''
+
+        ranked_df = self.items.sort_values(['ratings_ave', 'ratings_count'],
+                                           ascending=False)
+        ranked_df = ranked_df[ranked_df['ratings_count'] > 3]
```

Comment:
`3` should be a parameter

Reply:
https://github.com/scikit-learn/scikit-learn/blob/7813f7efb/sklearn/ensemble/forest.py#L758